Upgrade to Groovy 5

bentsherman opened the pull request:
Initial attempt to upgrade to [Groovy 5](https://groovy-lang.org/releasenotes/groovy-5.0.html).

Benefits:
- new collection functions like `zip` (can be added to Nextflow standard library)
- pattern matching for `instanceof` (can be added to Nextflow language)
- use `_` for placeholder variables (can be added to Nextflow language)

Changes:

- Had to remove generic type arguments from `instanceof` checks as the Groovy parser doesn't allow it anymore

- Had to add `@Slf4j` to some inner classes to make `log` work in that class. not sure if this is a bug or just a breaking change 

Current issues:

- Groovy complained about a few implementers of `CacheFunnel`, even though I didn't see anything wrong. Converting CacheFunnel to Java fixed the issue for one but not the other

- Multiple issues with the type checker that I'm pretty sure are just Groovy bugs, particularly `CmdLog` and `DumpOp`

- The type checker does not recognize `target` in `TaskConfig` as a Map, seems like a bug. I wonder if it's an edge case since `target` is defined in super class as private and a Delegate

Overall it looks like Groovy 5 beta still has a few compiler bugs to sort out. But I'm excited to get it working so that I can bring the new pattern matching features into the Nextflow language.

bentsherman committed:
Upgrade to Groovy 5

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

pditommaso commented:
Worth following this discussion, it looks like there are not negligile perf issue with Groovy 5 and invoke dynamic https://github.com/apache/grails-core/issues/15293

bentsherman commented on modules/nextflow/src/main/groovy/nextflow/cli/CmdLog.groovy:
```diff
@@ -299,7 +300,8 @@ class CmdLog extends CmdBase implements CacheBase {
                     String line
-                    while( (line=reader.readLine()) && c++<MAX_LINES ) {
+                    // ERROR: Cannot call java.lang.Integer#compareTo(java.lang.Integer) with arguments [java.lang.Object]
+                    while( (line=reader.readLine()) && c++ < MAX_LINES ) {
```
This seems to be caused by the presence of the delegate field, which can obscure other fields. I have seen issues like this before when using delegates. I submitted a ticket to the Groovy maintainers, but for now we can just do `(int) MAX_LINES` and move on.

bentsherman committed:
Preserve Groovy 4 file/path truthiness

Groovy 5 changes File/Path truthiness so that `if (file)` evaluates to
`file.exists()` rather than `file != null` (GROOVY-11996). Nextflow relies
on the legacy semantics where any non-null File/Path is truthy, so set the
`groovy.truth.file.exists.enabled=false` system property (added in Groovy
5.0.7) across all JVM launch paths to restore the previous behavior.

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
Groovy 5.0.7 released, though still waiting for Maven Central to be updated

Applied separate fix for Path truthiness. Will re-run CI when Maven is ready

bentsherman commented:
Relevant Groovy 5 commit: https://github.com/apache/groovy/commit/a1c006c974a3aca7f3d8c37c607535849e388580

bentsherman committed:
Fix Groovy 5 runtime behavior changes

Address several runtime failures introduced by the Groovy 5 upgrade:

- ScriptCompiler: pass the class simple name as the import alias instead
  of null. Groovy 5's ModuleNode.checkUsage throws a NPE for imports with
  a null alias, which broke compilation of virtually every pipeline script.

- TaskConfig: add a setProperty override that routes to put(). Groovy 5
  now throws ReadOnlyPropertyException when assigning a property that only
  has a getter (e.g. `config.shell = ...`); Groovy 4 fell through to Map.

- scm repository providers: drop the method-level generic signatures from
  `invokeAndResponseWithPaging`. Groovy 5 eagerly introspects the class via
  java.beans.Introspector on construction, which throws a NPE in the JDK
  TypeResolver for methods with method-level type variables.

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix Groovy 5 test failures: ProcessConfig property and void mock stubs

- ProcessConfig: add setProperty override routing to put(), mirroring
  TaskConfig, so directive assignment does not throw ReadOnlyPropertyException.

- CondaCacheTest: stub the int-returning runCommand mock with `>> 0` instead
  of `>> null` (Groovy 5 no longer coerces null to primitive int).

- MailerTest, SpackCacheTest, LocalPollingMonitorTest: stub void mocked
  methods with `>> { }` instead of `>> null` (Groovy 5 rejects casting null
  to void).

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix Groovy 5 input resolution: InParam index/mapIndex constants

Under Groovy 5, a bare typed field in an interface (e.g. `short index`)
compiles to a `static final` constant (value 0) instead of an abstract
property. That constant shadowed the real instance value during dynamic
property access in `BaseInParam.decodeInputs`, causing `mapIndex` to read
as 0 instead of -1 and taking the wrong decode branch -- returning the
first element of a list value instead of the value itself.

Declare `getIndex()`/`getMapIndex()` as explicit abstract getters (matching
the interface's existing getter style) and implement them in the v2
`ProcessInput`. Fixes TaskProcessorTest input file arity failures.

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix Groovy 5 TaskConfig property routing and void mock stubs

TaskConfig.setProperty: only route to `put` when there is no real setter.
The previous unconditional `put` clobbered real property setters (notably
`setContext`), leaving the dynamic-evaluation binding unset so directive
closures/GStrings resolved against the wrong scope. Also call `setBinding`
explicitly from `setContext` instead of assigning the LazyMap private field
via a bare property (which the override would intercept).

Tests: replace `>> null` on void Spy/Mock methods with `>> { }` (or drop it)
— under Groovy 5 a constant `null` response cannot be cast to void.
Covers TaskRunTest, SplitOpTest, ReportObserverTest, TaskPollingMonitorTest.

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix Groovy 5 splitter tests: use concrete subclass over map coercion

Under Groovy 5, `[:] as AbstractSplitter` produces a proxy that no longer
delegates the abstract class's concrete methods (invokeEachClosure,
findSource, getCollectorBaseFile), so calls fail with MissingMethodException.
Replace the map coercions with small concrete test subclasses.

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix Groovy 5: set ext binding via explicit setter; int mock stub

- TaskConfig.setContext: call ext.setBinding(context) explicitly so the
  'ext' LazyMap's dynamic-evaluation binding is set (a bare property
  assignment is not routed to the protected setter under Groovy 5),
  fixing dynamic ext directive resolution (TaskHasherTest).
- TaskHandlerTest: map the null maxForks case to 0 — a null constant
  cannot be cast to the primitive int return under Groovy 5.

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix scan accumulator regression and hermetic jgit commits

- IterableDef.accumulator: the Groovy 5 rewrite of the mapper stopped
  mutating the accumulator list, so each scan iteration only fed back the
  most recent output instead of all previous ones (ScriptRecurseTest
  'should recurse with scan' produced 63 instead of 74). Restore the
  accumulate-then-snapshot semantics.
- MultiRevisionRepositoryStrategyTest: disable commit signing in the jgit
  commits so the test does not depend on the developer's global
  commit.gpgsign git config.

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix Groovy 5 plugin test mock stubs (void/int)

Under Groovy 5 a constant `>> null` response cannot be cast to a void or
primitive-int method. Replace with `>> { }` for void Spy/Mock methods (or
map null->0 for the int case):
- nf-amazon AwsBatchTaskHandlerTest: getMaxTransferAttempts (int), terminateJob (void)
- nf-azure AzBatchServiceTest/AzBatchTaskHandlerTest: terminateJobs/cleanupJobs/
  cleanupPools/createPool/deleteTask (void)
- nf-google GoogleBatchTaskHandlerTest: deleteJob (void)
- nf-k8s K8sClientTest/K8sDriverLauncherTest/K8sTaskHandlerTest: setupHttpsConn/
  setRequestMethod/createK8sConfigMap/waitPodStart/printK8sPodOutput/
  deleteConfigMap/saveJobLogOnError/deleteJobIfSuccessful (void)

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix Groovy 5 nf-wave test void mock stub

setDisableRemoteBinDir is a void setter; a `>> null` constant response
cannot be cast to void under Groovy 5. Drop the response in WaveFactoryTest.

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix Groovy 5 static-compiler field-type narrowing in AssetManager

Under Groovy 5, `provider instanceof LocalRepositoryProvider` applied
directly to the `provider` field caused @CompileStatic to narrow the field's
inferred type to LocalRepositoryProvider class-wide. This inserted bogus
`checkcast LocalRepositoryProvider` instructions in clone(),
getRemoteRevisions() and updateModules(), which threw ClassCastException at
runtime for any non-local provider (e.g. GithubRepositoryProvider) -- breaking
AssetManagerTest.testClone and CmdCloneTest.testClone.

Assign the field to a local variable before the instanceof check so the
narrowing applies to the local, not the field.

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented on modules/nextflow/src/main/groovy/nextflow/scm/AssetManager.groovy:
```diff
@@ -659,7 +659,13 @@ class AssetManager implements Closeable {
      */
     boolean isLocalScmSource() {
-        return provider instanceof LocalRepositoryProvider
+        // assign to a local before the `instanceof` check: under Groovy 5 the
+        // static compiler otherwise narrows the `provider` *field* type to
+        // LocalRepositoryProvider class-wide, inserting bogus casts in other
+        // methods (e.g. clone/getRemoteRevisions/updateModules) that fail with
+        // ClassCastException for non-local providers.
+        final p = provider
+        return p instanceof LocalRepositoryProvider
```
Looks like a Groovy 5 bug, I will submit a bug report

bentsherman replied:
https://issues.apache.org/jira/browse/GROOVY-12166

bentsherman commented:
Tests are finally passing!

TODO:
- [ ] Remove local var + instanceof workaround after Groovy bug is fixed
- [ ] Review changes to TaskConfig carefully
